src/file/writer.rs

```diff
@@ -15,16 +15,290 @@
+    ));
+
+    Ok(self.row_group_writer.as_mut().unwrap())
+  }
+
+  fn close(&mut self) -> Result<()> {
+    if self.is_closed {
+      return Err(general_err!("File writer is closed"));
+    }
+    self.is_closed = true;
+    self.finalise_row_group_writer()?;
+    self.write_metadata()
+  }
+}
+
+/// Serialized row group writer.
+///
+/// Coordinates writing of a row group with column writers.
+pub struct SerializedRowGroupWriter {
+  descr: SchemaDescPtr,
+  props: WriterPropertiesPtr,
+  file: File,
+  total_rows_written: Option<u64>,
+  total_bytes_written: u64,
+  column_index: usize,
+  current_column_writer: Option<ColumnWriter>,
+  row_group_metadata: Option<RowGroupMetaDataPtr>,
+  column_chunks: Vec<ColumnChunkMetaDataPtr>
+}
+
+impl SerializedRowGroupWriter {
+  pub fn new(
+    schema_descr: SchemaDescPtr,
+    properties: WriterPropertiesPtr,
+    file: &File
+  ) -> Self {
+    let num_columns = schema_descr.num_columns();
+    Self {
+      descr: schema_descr,
+      props: properties,
+      file: file.try_clone().unwrap(),
+      total_rows_written: None,
+      total_bytes_written: 0,
+      column_index: 0,
+      current_column_writer: None,
+      row_group_metadata: None,
+      column_chunks: Vec::with_capacity(num_columns)
+    }
+  }
+
+  /// Checks and finalises current column writer.
+  fn finalise_column_writer(&mut self) -> Result<()> {
+    if let Some(ref mut writer) = self.current_column_writer {
+      let (bytes_written, rows_written, metadata) = match writer {
+        ColumnWriter::BoolColumnWriter(typed) => typed.close()?,
+        ColumnWriter::Int32ColumnWriter(typed) => typed.close()?,
+        ColumnWriter::Int64ColumnWriter(typed) => typed.close()?,
+        ColumnWriter::Int96ColumnWriter(typed) => typed.close()?,
+        ColumnWriter::FloatColumnWriter(typed) => typed.close()?,
+        ColumnWriter::DoubleColumnWriter(typed) => typed.close()?,
+        ColumnWriter::ByteArrayColumnWriter(typed) => typed.close()?,
+        ColumnWriter::FixedLenByteArrayColumnWriter(typed) => typed.close()?
+      };
+
+      // Update row group writer metrics
+      self.total_bytes_written += bytes_written;
+      self.column_chunks.push(metadata);
+      if let Some(rows) = self.total_rows_written {
+        if rows != rows_written {
+          return Err(general_err!(
+            "Incorrect number of rows, expected {} != {} rows",
+            rows,
+            rows_written
+          ));
+        }
+      } else {
+        self.total_rows_written = Some(rows_written);
+      }
+    }
+    Ok(())
+  }
+}
+
+impl RowGroupWriter for SerializedRowGroupWriter {
+  #[inline]
+  fn num_columns(&self) -> usize {
+    self.descr.num_columns()
+  }
+
+  #[inline]
+  fn has_next_column(&self) -> bool {
+    self.column_index < self.num_columns() && self.row_group_metadata.is_none()
+  }
+
+  #[inline]
+  fn next_column(&mut self) -> Result<&mut ColumnWriter> {
+    if self.row_group_metadata.is_some() {
```

Comment:
nit: we can just call `has_next_column` at the beginning.

Reply:
I am going to make it iterator-like, this condition will be updated.